### DISCUSSION TITLE
State changes when no reducer perform changes

### DISCUSSION THREAD

**Discussion by BrainCrumbz** [reduxjs/redux-toolkit#2933] [Q&A] [answered]
Hi all. Not sure if the title conveys its intended meaning, please bear with us.

In a plain redux app (without redux-toolkit), when you dispatch an action and no reducer acts on that, then no change of state is notified. See console logs in [this stackblitz](https://stackblitz.com/edit/redux-example-101-jmmbsi?file=index.js). (Dependencies: redux 4.2.0)

In a redux-toolkit app, in the same scenario, state changes are notified upon any dispatch, even if the state is not really changed. See console logs in [this stackblitz](https://stackblitz.com/edit/react-nwus7m?file=src/App.js). (Dependencies: @reduxjs/toolkit 1.9.0)

Of course, it does not make sense to dispatch an action not mapped in any reducer. But this could simulate the case where a reducer maps on an action, but then no real state change is made.

Is this behaviour intended? Does it cause any harm, e.g. to performance?

Just to give some context: we are working on a custom store enhancer, which adds some info to the store, and at some point we were noticing state change notification even when nothing was really changing. So we started wondering whether it was our enhancer or what.

Also, please point us in the right direction if this is not the right place for this topic.  
Thanks all

**Comment by markerikson** [comment, marked as answer]
Can you clarify what those two examples are trying to show?  They appear to be very different setups.

For the record, the core Redux store _always_ runs _all_ subscriber callbacks (ie calls to `store.subscribe(callback)`) after _every_ action, regardless of whether the state was updated or not.  RTK just ends up using `createStore` internally, so there is no difference in behavior between "vanilla Redux" and "RTK" here.

**Reply by BrainCrumbz** [reply]
The idea - sorry if it was not clear - is to have two dummy "counter" example apps. One using "raw" redux, the other using RTK. In both apps, the dispatched action is **not** caught by the (only) reducer slice. In both apps, the store is subscribed to in order to log its state.

Actually, in raw-redux app the the store is not subscribed to, but instead state is logged within `mapStateToProps`. Not sure if this makes a difference. From your answer, is this the part causing an apparently different behaviour?

**Reply by BrainCrumbz** [reply]
Ok, that was it. Those were two example stackblitz modified to the purpose, that's why the different setup. Thanks for pointing out the difference. 

Should this be deleted? Or closed?

**Reply by markerikson** [reply]
Nothing wrong with leaving it open :)

And yeah, those _are_ two very different aspects.

If you add a `store.subscribe()` call in both examples, you'll see that in _both_ of those the subscriber callback _always_ runs after _every_ dispatch.

React-Redux itself calls `store.subscribe()` internally.

**HOWEVER**: React-Redux specifically implements internal optimizations and checks to see if `store.getState() === lastStoreState`.  If the state has not changed, then it immediately bails out without even trying to run your `mapState` or `useSelector` functions.  If the store state _has_ changed, it will run those selectors, and then diff the results to see if this component needs to re-render.  So, there's multiple layers of behavior here:

- The store _always_ notifies _all_ subscribers
- In a normal app there really aren't any manual `store.subscribe()` calls - the only `store.subscribe()` calls are internal to React-Redux
- React-Redux checks to see if the root state changed at all, and assumes that if it's the same reference that no other nested values changed (because an immutable update requires changing all nested object references)
- If the root state _did_ change it checks on a per-component basis to see what components now need updated data from the store

Some references that might be helpful:

- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/
- https://blog.isquaredsoftware.com/2018/11/react-redux-history-implementation/
- https://blog.isquaredsoftware.com/2019/06/presentation-react-redux-deep-dive/

**Reply by markerikson** [reply]
@BrainCrumbz : purely out of curiosity, what sort of a custom store enhancer are you working on, and how is it "adding info to the store"?  Always interested in hearing details on what people are doing with Redux?

**Reply by BrainCrumbz** [reply]
Thanks for the details on what's happening under the hood. Some of them were known, but not all.

Oh, and there's an typo from our side: adding info to the _state_. Not the store.

Custom enhancer is our take at "return side effect from reducer". We've been reading through various issue threads about the topic, about the new listener middleware, what are the possible approach to side-effects with/besides RTK. We're fiddling with this custom enhancer as we'd like to try this scheme:

1. UI or multimedia component dispatches action
2. reducer performs its logic and decides whether some side-effect is needed (mostly on multimedia, not HTTP calls), so it returns some command
3. command listeners implement side-effect

This is similar to what redux-loop does, but we don't want to "pack" the functions to be called within the "command" (_loop_ for them). It is similar to what listener middleware and redux-observable and others do, with the main difference that side-effects do not wait for actions (generated by UI and such) but for commands (generated, if any, from reducer logic).

To us, the main difference on the approach lies here: 
* in our type of app deciding whether to issue a side-effect or not requires some logic, it's not a straightforward consequence of some UI event happening. 
* If side-effects listen to actions, then this logic must be put in side-effects, and probably part of that logic will be needed in reducer as well, so there will be overlapping or splitting between the two.
* If side-effects listen to commands, the this logic is all in one place within the reducer, bringing a number of advantages.

I mean, it's not that is not doable with action-based effects, but...  
BTW this app _could_ potentially be redux-only, i.e. no react on top.

**Reply by markerikson** [reply]
Ah, interesting! Haven't seen anyone try that in a while.

**Reply by BrainCrumbz** [reply]
Yeah, it's somehow like "ok, now's their turn. Let's see what happens" :)

One PoC we'll make is centered around the "user interaction token". There's a better term for that and you will already know it, but my google-skills are failing me now. It's that feature for which some browser APIs are allowed only when called *strictly within* a user event handler. Some multimedia APIs fall under this category.

So, for example's sake, say that `HTMLMediaElement.play()` or `HTMLMediaElement.load()` is one of those, and this is considered as a side-effect called in some listener. If side-effect listens to actions, we have:

```
UI event handler -> dispatch -> reducer -> listener middleware (or other) -> action listener -> htmlMediaElem.play()
```

If side-effect listens to commands put in place by our new enhancer (with its meta-reducer and the logic to invoke registered listeners), we have:

```
UI event handler -> dispatch -> reducer -> our custom enhancer -> command listener -> htmlMediaElem.play()
```

In any case, for this to work, the whole chain must be called without returning control from the UI event handler to the runtime. If e.g. our enhancer calls command listeners asynchronously, after resolving some promise, this should _not_ work.

**Reply by BrainCrumbz** [reply]
Hi there. Just for completeness: PoC with our custom enhancer _works_, i.e. the call chain is synchronous all the way down to command listener. We'd expect the same happens with middleware enhancer, in case we'd decide to use it.

**Reply by markerikson** [reply]
I'd be interested in seeing an example app and the actual enhancer implementation if you get a chance!

**Reply by BrainCrumbz** [reply]
Cool! You should have received an invitation to a private repo.
